avro4s-core/src/main/scala/com/sksamuel/avro4s/DefinitionEnvironment.scala

```diff
@@ -0,0 +1,29 @@
+package com.sksamuel.avro4s
+
+import scala.reflect.runtime.universe._
+
+/**
+ * Helps with encoders/decoders/schemas for recursive types. Provides a map of already defined bindings for types.
+ * For an Encoder e.g., captures which encoders are already defined when building the (cyclic) encoder graph for
+ * a recursive type.
+ *
+ * @tparam Typeclass Encoder, Decoder or SchemaFor
+ */
+class DefinitionEnvironment[Typeclass[_]](
```

Comment:
DefinitionEnvironment is used keeps a map of typess we have already seen and the typeclasses returned for those?Magnolia is meant to handle recursive types for us, does this fail ?

Reply:
One complication I see when using Magnolia's capabilities for recursive encoders and decoders is that the caching of name mappings would need to introduce synchronization locks to be correct.

Reply:
Thank you that's perfect. Would you mind adding what you wrote as documentation onto the def environment so it's saved for future readers.